bundles/org.openhab.binding.mqtt.ruuvigateway/src/main/java/org/openhab/binding/mqtt/ruuvigateway/internal/handler/RuuviTagHandler.java

```diff
@@ -0,0 +1,447 @@
+            if (cacheHint == null || cacheHint instanceof Unit<?>) {
+                Unit<?> unit = (Unit<?>) cacheHint;
+                initNumberStateCache(channelUID, unit);
+            } else {
+                Class<?> cacheType = (Class<?>) cacheHint;
+                initCacheWithClass(channelUID, cacheType);
+            }
+
+        }
+    }
+
+    private <T extends Quantity<T>> RuuviCachedNumberState<?> initNumberStateCache(ChannelUID channelUID,
+            @Nullable Unit<T> unit) {
+        final RuuviCachedNumberState<?> cached;
+        if (unit == null) {
+            cached = new RuuviCachedNumberState<>(channelUID);
+            channelStateByChannelUID.put(channelUID, cached);
+        } else {
+            cached = new RuuviCachedNumberState<>(channelUID, unit);
+            channelStateByChannelUID.put(channelUID, cached);
+        }
+        return cached;
+    }
+
+    private ChannelState initCacheWithClass(ChannelUID channelUID, Class<?> clazz) {
+        try {
+            ChannelState cached = (ChannelState) clazz.getConstructor(ChannelUID.class).newInstance(channelUID);
+            Objects.requireNonNull(cached); // to make compiler happy
+            channelStateByChannelUID.put(channelUID, cached);
+            return cached;
+        } catch (InstantiationException | IllegalAccessException | IllegalArgumentException | InvocationTargetException
+                | NoSuchMethodException | SecurityException e) {
+            throw new IllegalStateException(e);
+        }
+    }
+
+    @Override
+    protected CompletableFuture<@Nullable Void> start(MqttBrokerConnection connection) {
+        if (topic == null) {
+            // Initialization has not been completed successfully, return early without changing
+            // thing status
+            return CompletableFuture.completedFuture(null);
+        }
+
+        updateStatus(ThingStatus.UNKNOWN);
+        return connection.subscribe(topic, this).handle((subscriptionSuccess, subscriptionException) -> {
+            if (subscriptionSuccess) {
+                updateStatus(ThingStatus.ONLINE, ThingStatusDetail.NONE, "@text/online.waiting_initial_data");
+                heartbeatFuture = scheduler.scheduleWithFixedDelay(this::heartbeat, heartbeatTimeoutMillisecs,
+                        heartbeatTimeoutMillisecs, TimeUnit.MILLISECONDS);
+            } else {
+                if (subscriptionException == null) {
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                            "MQTT subscription failed");
+                } else {
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                            getTranslated("offline.communication-error.mqtt_subscription_failed",
+                                    new Object[] { subscriptionException.getClass().getSimpleName(),
+                                            subscriptionException.getMessage() }));
```

Comment:
I think this could be simplified. I'm not sure about the syntax for multiple parameters, but as an example:
https://github.com/openhab/openhab-addons/blob/f2e7597db69598b40b7f9bae24c7ac5dedabb345/bundles/org.openhab.binding.danfossairunit/src/main/java/org/openhab/binding/danfossairunit/internal/handler/DanfossAirUnitHandler.java#L107-L108

with corresponding text:
https://github.com/openhab/openhab-addons/blob/f2e7597db69598b40b7f9bae24c7ac5dedabb345/bundles/org.openhab.binding.danfossairunit/src/main/resources/OH-INF/i18n/danfossairunit.properties#L88

Reply:
Good tip! I assume it works similarly...